### ISSUE TITLE
Add support to SQLAlchemy 2.0

### ISSUE THREAD

**Issue by josecsotomorales** [kvesteri/sqlalchemy-continuum#326]

**Comment by zikphil** [comment]
Any ideas of the scope of work required for this and/or when it will be tackled? :)

**Comment by Jerommeke4** [comment]
There is another issue with the polymorphic_on check in `model_builder.py` which is not able to work with MappedColumn types. The following diff will support this.

```
--- a/sqlalchemy_continuum/model_builder.py
+++ b/sqlalchemy_continuum/model_builder.py
@@ -2,7 +2,7 @@ from copy import copy
 import six
 import sqlalchemy as sa
 from sqlalchemy.ext.declarative import declared_attr
-from sqlalchemy.orm import column_property
+from sqlalchemy.orm import column_property, MappedColumn
 from sqlalchemy_utils.functions import get_declarative_base
 
 from .utils import adapt_columns, option
@@ -78,6 +78,8 @@ def copy_mapper_args(model):
             column = model.__mapper_args__['polymorphic_on']
             if isinstance(column, six.string_types):
                 args['polymorphic_on'] = column
+            elif isinstance(column, MappedColumn):
+                args['polymorphic_on'] = column.column.key
             else:
                 args['polymorphic_on'] = column.key
     return args
```

**Comment by AlTosterino** [comment]
Bump 💪

**Comment by marksteward** [comment]
Hi sorry, will look at this shortly

**Comment by POD666** [comment]
Bump 🙂

**Comment by alliefitter** [comment]
Got [a PR up](https://github.com/kvesteri/sqlalchemy-continuum/pull/330) to address this. Dealing with a few broken tests and some hanging with postgres. I don't have much experience with sqlalchemy, so any help figuring out [the problem](https://github.com/kvesteri/sqlalchemy-continuum/pull/330#issuecomment-1509322077) would be greatly appreciated.

**Comment by marksteward** [comment]
Hi all, thanks very much for your help. I'm working through the suggested changes and the other issues that are currently causing tests to break but it's taken longer than expected.

In the meantime, I've raised https://github.com/kvesteri/sqlalchemy-continuum/issues/332 in case anyone has opinions on that.

**Comment by marksteward** [comment]
Just to update here, I'm having trouble with the postgres dialect (particularly the new [executemany code](https://github.com/kvesteri/sqlalchemy-continuum/blob/a4ca8bcb/sqlalchemy_continuum/manager.py#L459-L471), which now appears to be a single query), so I'll have to park this for a few days.

**Comment by marksteward** [comment]
Right, I've pushed to the [sqla-2.0](https://github.com/kvesteri/sqlalchemy-continuum/tree/sqla-2.0) branch and tests are passing except for the native ones. Is anyone able to have a look through for any issues?

This version only supports SQLAlchemy >= 1.4, but I figure that's reasonable as it's the transitional version and 1.3 is EOL.

Still to do:
- fix native versioning
- get rid of anything commented "< 1.4"
- get rid of six
- get rid of any more easy warnings
- see if we can get rid of some other pytest.skip()s
- update the README, CHANGES

I'll also be renaming the master branch to main with this release, just so people are aware.

**Comment by josecsotomorales** [comment]
Do we have any updates on this?

**Comment by marksteward** [comment]
I was hoping someone would be able to give it a test with their code, but failing that I'll probably do a release next week.

**Comment by josecsotomorales** [comment]
@marksteward I'm happy to help in whatever is needed to get this released, just let me know.

**Comment by josecsotomorales** [comment]
FYI tested the WIP branch in a personal project with SA and PostgreSQL, seems to be working fine.

**Comment by JPereira-FJB** [comment]
Hi. Do we have any updates?
Does SQLAlchemy-Continuum support SQLAlchemy 2.0 in any of the branches?

**Comment by josecsotomorales** [comment]
I've tested on this branch: https://github.com/kvesteri/sqlalchemy-continuum/tree/sqla-2.0 ... seems to be working fine for PostgreSQL. @marksteward what's needed to release this?

**Comment by josecsotomorales** [comment]
I'm happy to contribute to whatever we need to get this done.

**Comment by marksteward** [comment]
Thanks! I'm about to cut a 1.4 release, but if you want to work out any outstanding 2.0 warnings that'd be great.

Also, I've just realised I don't have permission to rename master to main. @kvesteri can you do that?

**Comment by marksteward** [comment]
OK, I've published 1.4. Please give it a go!

**Comment by josecsotomorales** [comment]
Thanks @marksteward! Will keep testing now with this release
